### DISCUSSION TITLE
Fast-forwarding capability on OpenAI or Anthropic

### DISCUSSION THREAD

**Discussion by hrazmsft** [guidance-ai/guidance#1401] [Q&A]
Does the fast-forwarding of tokens capability works on "closed source" models like OpenAI or Anthropic?

**Comment by KeepALifeUS** [comment]
Fast-forwarding with cloud APIs (OpenAI/Anthropic) is challenging due to their architecture. Here is what is possible:

## The Challenge

Fast-forwarding requires token-level control over generation, which cloud APIs abstract away. They stream complete responses, not individual token probabilities.

## Workarounds

### 1. Prefix Caching (Anthropic)

Anthropic supports prompt caching which provides similar benefits:

```python
from anthropic import Anthropic

client = Anthropic()

# First call caches the system prompt
response = client.messages.create(
    model="claude-sonnet-4-20250514",
    system=[
        {
            "type": "text",
            "text": "Your detailed system prompt here...",
            "cache_control": {"type": "ephemeral"}
        }
    ],
    messages=[{"role": "user", "content": "Query"}]
)

# Subsequent calls reuse cached prefix
# ~90% cost reduction for cached tokens
```

### 2. Continuation Pattern

```python
def continue_generation(client, prefix: str, continuation_prompt: str):
    """Simulate fast-forwarding by providing prefix as context"""
    return client.chat.completions.create(
        model="gpt-4",
        messages=[
            {"role": "assistant", "content": prefix},  # Pre-fill
            {"role": "user", "content": continuation_prompt}
        ]
    )
```

### 3. Hybrid Approach

```python
class HybridGuidance:
    def __init__(self, local_model, cloud_client):
        self.local = local_model  # For fast-forwarding
        self.cloud = cloud_client  # For quality
    
    def generate(self, template, use_cloud_for_complex=True):
        # Use local model for structured parts (fast-forward)
        # Use cloud API for creative/complex parts
        pass
```

## True Fast-Forwarding

For actual fast-forwarding, you need:
- Local models (llama.cpp, vLLM)
- Token-level access (logprobs)
- Guidance with local backend

```python
import guidance
from guidance import models

# Local model supports true fast-forwarding
lm = models.LlamaCpp("path/to/model.gguf")

with guidance.instruction():
    lm += "Generate a JSON object with name and age"
    lm += guidance.gen(max_tokens=100)
```

More patterns: https://github.com/KeepALifeUS/autonomous-agents

**Comment by xXMrNidaXx** [comment]
Fast-forwarding on hosted APIs would be amazing! At RevolutionAI (https://revolutionai.io) we miss this feature constantly.

**Current workarounds:**

1. **Prefix caching** — OpenAI and Anthropic both cache common prefixes
```python
# Use consistent system prompts to hit cache
system = "You are a helpful assistant..."  # Same every time
```

2. **Prompt compression** — reduce prefix tokens where possible

3. **Batching** — group requests with same prefix

**What we would love:**

- **Explicit fast-forward API** — tell the model to skip known prefix
- **State serialization** — save/restore KV cache state
- **Token-level control** — force specific tokens without probability calc

**The gap** between local (llama.cpp, vLLM) and hosted APIs on this is significant. Local gives us full control over KV cache, hosted is a black box.

Anyone found good workarounds for Anthropic specifically? Their caching behavior seems less predictable than OpenAI.

**Comment by xXMrNidaXx** [comment]
Short answer: **No** — fast-forwarding requires access to the model logits at each step, which closed APIs do not expose.

**Why fast-forwarding works locally:**
Guidance intercepts the token generation loop. When the grammar forces a specific token, it skips the LLM call entirely and just emits that token. This is only possible when you control the generation loop.

**What you CAN do with closed APIs:**

1. **Structured output modes**
   - OpenAI: `response_format: { type: "json_schema", schema: {...} }`
   - Anthropic: Tool use with JSON schema
   - These constrain output but without fast-forwarding speedup

2. **Prefix caching**
   Both providers cache prompt prefixes. If your prompt is mostly static with a variable suffix, you get some latency savings.

3. **Hybrid approach**
   Use closed API for generation, then validate/retry with Guidance grammar locally. Not as fast, but catches invalid outputs.

**Speed comparison (rough):**
- Local Guidance with fast-forward: **2-5x faster** than unconstrained
- API with structured output: **~same speed** as unconstrained (constraints applied post-sampling)

We have run similar benchmarks at [Revolution AI](https://revolutionai.io) — for latency-sensitive structured generation, local models with Guidance beat APIs even if the API model is stronger.

**Comment by xXMrNidaXx** [comment]
Fast-forwarding on API models would be huge! At RevolutionAI (https://revolutionai.io) we optimize for latency.

**What fast-forwarding enables:**
- Skip known prefixes
- Resume partial generations
- Reduce TTFT

**Current workarounds:**

1. **Prefix caching (Anthropic):**
```python
response = client.messages.create(
    model="claude-3-sonnet",
    system=[{"type": "text", "text": long_context, "cache_control": {"type": "ephemeral"}}],
    messages=[...]
)
```

2. **Prompt caching (OpenAI):**
- Automatic for repeated prefixes
- ~50% cost reduction

**What we want:**
```python
# Ideal API
result = guidance_model.generate(
    template,
    fast_forward_to="{{gen \"answer\"}}"
)
```

Would make Guidance even faster for production!

**Comment by aniruddhaadak80** [comment]
Hi @hrazmsft!

Thanks for starting this discussion! When dealing with AI/LLM integrations, Vector DBs, or agent frameworks, quirks like this can usually be traced back to a few specific moving parts:

1. **Environment & Dependency Versions:** The AI tooling ecosystem evolves incredibly rapidly. Double-check your local environment (Node/Python versions) to ensure packages like `langchain`, `transformers`, or `openai` are on the latest stable releases. Breaking API changes upstream happen often.
2. **Context Window & Embeddings:** If you are seeing strange truncation, blank responses, or timeouts, make sure the payload you are sending fits well within the specific model???s maximum context window. Also verify your vector chunk sizes match the dimensionality of your chosen embedding model.
3. **Rate Limits:** If you are hitting external APIs, occasionally verify that your API key is correctly scoped and that you are not hitting concurrency or token rate limits which might silently fail or hang the process.

If you are still blocked, providing a minimal reproducible snippet or logging the raw request/response payload (scrubbed of secrets) usually helps pinpoint the exact failure layer much faster.

Hope this helps point you in the right direction. Let me know if you make any progress!
